### PR TITLE
Implemented logger that supports ITM, RTT, and semihosting

### LINKED ISSUE #7: Examples use ITM
May cause a hang when running examples, if debugger is removed or mis-configured.

Feature gating ITM usage is a possible solution.

### LINKED ISSUE #104: [Proposal] Log wrapper
For my own project I wrapped the log crate to allow output via RTT or none via feature. A feature gated logger implementing several implementations sounds really useful. Could include ITM, RTT, etc. Leaving no features selected makes the macro empty and it get's compiled out.

Is this the right level of crate for this or should it go somewhere else?

[Example](https://gist.github.com/mtthw-meyer/0c1d4977cfbb848916f92830c35c852c)

### PR DESCRIPTION
**Pull request by mtthw-meyer**
Implemented logger that supports ITM, RTT, and semihosting.
Updated examples to use the log crate.
I have tested it with RTT and none.
Closes #104 
Closes #7

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -30,15 +30,23 @@ void = { version = "1.0.2", default-features = false }
 cast = { version = "0.2.3", default-features = false }
 nb = "0.1.2"
 paste = "0.1.18"
+log = "0.4.11"
+panic-halt = "0.2.0"
```
Comment (hargoniX) [inline Cargo.toml:34]: Does panic-halt really have to be in our default enabled dependencies? I personally don't like the idea of having a default panic implementation as a HAL when usually the user wants to determine that implementation in their crate themselves.

File: src/logger.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                inner: semihosting::InterruptOk::<_>::stdout().expect("Get Semihosting stdout"),
+            };
+        }
+
+        pub fn init() {
+            cortex_m_log::log::init(&LOGGER).unwrap();
+        }
+    }
+    else {
+        use panic_halt as _;
```
Comment (hargoniX) [inline src/logger.rs:84]: See my first comment for this.

File: src/logger.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+cfg_if::cfg_if! {
```
Comment (hargoniX) [inline src/logger.rs:1]: I also do not see anything stm32h7 specific in this file (which is the only file that actually contains library code and not examples as far as I can see? ) so I don't believe this should live in stm32h7xx-hal, after all we are building a Hardware Abstraction Layer and this code does not exactly abstract hardware away, it provides some minimal logic around code that is already general across Cortex-M chips.

File: examples/rtic.rs [hunk 4]
```diff
@@ -3,7 +3,6 @@
 #![no_std]
 #![no_main]
 
-extern crate panic_itm;
```
Comment (richardeoin) [inline examples/rtic.rs:6]: Every examples needs a `#[panic_handler]` to compile. For the examples that don't use logging, then I'd say `use panic_halt as _;` is fine

File: Cargo.toml (hunk 1 above)
Comment (richardeoin) [reply Cargo.toml:34]: Agreed, having panic implementation as a required dependency is not helpful

**Comment by richardeoin** [comment]
I really like using the `log` crate in the examples, because log works across so much of the rust ecosystem. And the example are simpler + more flexible! However having the log implementation as part of the crate is a problem as @hargoniX highlighted.

That can be easily solved by moving `src/logger.rs` -> `examples/logger.rs`, and replacing `use stm32h7xx_hal::logger;` with `mod logger;`. You'd need to document that it's not a standalone example, and tell cargo not to try to build it individually. This should work:

```
[[example]]
name = "logger"
required-features = ["!"] # logger is used by the other examples, not a standalone example
```

Or even move it to `examples/utilities/logger.rs` (for example) to make clear what it is.

Then all the added dependencies can actually be dev-dependencies instead.

**Comment by mtthw-meyer** [comment]
Sure, I can shuffle things around to make that work.

**Comment by mtthw-meyer** [comment]
Yes, `log` is practically a std library.

For building examples I see two options.

1. Leave in features for logging that only effect the examples. Downside is this recompiles the entire crate when any of the features are changed. Upsides are general compatibility with other tools and ease of use.

2. Use cfg flags which require each example to be compiled separately with "cargo rustc --features=stm32h750v,rt --examples -- --cfg log_rtt" and do not work with other tools (e.g. cargo embed can't handle -- --cfg but works fine with --features). I think this only works with cargo rustc invocations.. Advantage being the crate doesn't get recompiled.

Given the limitations of the 2nd approach I favor the 1st. Cargo simply isn't good at handling passing around a bunch of cfg flags yet. Unless there is another approach I am unaware of.

**Comment by richardeoin** [comment]
LGTM! I think the 1st option is the best approach too. It is a bit awkward to have feature flags that don't affect the crate source itself, but it wins on compatibility/ease of use.

The only other thing it would be nice to add is more documentation in `examples/README.md`

**Comment by mtthw-meyer** [comment]
The ITM and semishosting still need to be tested.

**Comment by richardeoin** [comment]
I tested semihosting and ITM, they both work. I've added some more docs for ITM.

**Comment by richardeoin** [comment]
Rebased onto master

**Comment by mtthw-meyer** [comment]
Looks good!

**Comment by richardeoin** [comment]
I fixed the recently added examples that weren't using the logger

**Comment by richardeoin** [comment]
Ok, let's merge this! Thanks for your work @mtthw-meyer 

bors r+

**Pull request merged**